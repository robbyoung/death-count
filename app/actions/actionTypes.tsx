export enum ActionType {
    ADD_PLAYTHROUGH = 'Add Playthrough',
    ADD_GAME = 'Add Game',
    ADD_DEATH = 'Add Death',
    ADD_DEATH_DETAIL = 'Add Death Detail',
    COMPLETE_DEATH = 'Complete Death',
    ADD_OPTION = 'Add Option',
    ADD_OPTION_SET = 'Add Option Set',
    SELECT_GAME = 'Select Game',
    LOAD_STATE = 'Load State',
    COMPLETE_GAME = 'Complete Game',
    SELECT_PLAYTHROUGH = 'Select Playthrough',
    SELECT_OPTION_SET = 'Select Option Set',
    DELETE_PLAYTHROUGH = 'Delete Playthrough',
}
