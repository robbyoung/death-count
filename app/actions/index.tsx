export * from './actionTypes';
export * from './addPlaythrough';
export * from './addDeath';
export * from './addGame';
export * from './addDeathDetail';
export * from './completeDeath';