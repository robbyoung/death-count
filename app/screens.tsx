export enum Screens {
    Home = 'Home',
    NewDeath = 'NewDeath',
    GamesList = 'GamesList',
}
