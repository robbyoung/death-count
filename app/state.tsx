export interface Game {
    id: string;
    name: string;
    options: string;
}

export interface DeathOptions {
    name: string;
    options: string[];
}

export interface Playthrough {
    id: string;
    name: string;
    description: string;
    gameId: string;
}

export interface Death {
    id: string;
    playthroughId: string;
}

export interface State {
    games: Game[];
    playthroughs: Playthrough[];
    deaths: Death[];
    currentPlaythrough: string;
    currentGame: string;
}