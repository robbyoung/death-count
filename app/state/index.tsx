export * from './state';
export * from './testState';