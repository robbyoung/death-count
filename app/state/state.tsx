export interface Game {
    id: string;
    name: string;
    selected: boolean;
    complete: boolean;
}

export interface Playthrough {
    id: string;
    name: string;
    gameId: string;
    selected: boolean;
}

export interface Death {
    id: string;
    playthroughId: string;
    complete: boolean;
    details: { [id: string] : string };
}

export interface OptionSet {
    id: string;
    gameId: string;
    title: string;
    options: string[];
    selected: boolean;
}

export interface State {
    games: Game[];
    playthroughs: Playthrough[];
    deaths: Death[];
    optionSets: OptionSet[];
}