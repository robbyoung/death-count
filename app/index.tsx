import { createAppContainer } from 'react-navigation';
import { createStackNavigator } from 'react-navigation-stack';
import { Screens } from './screens';
import Home from './containers/home';
import NewDeath from './containers/newDeath';
import GamesList from './containers/gamesList';
import NewGame from './containers/newGame';
import Loading from './containers/loading';
import Startup from './containers/startup';
import PlaythroughsList from './containers/playthroughsList';

const MainNavigator = createStackNavigator(
	{
		Loading: { screen: Loading },
		Home: { screen: Home },
		NewDeath: { screen: NewDeath },
		GamesList: { screen: GamesList },
		NewGame: { screen: NewGame },
		Startup: { screen: Startup },
		PlaythroughsList: { screen: PlaythroughsList }
	},
	{
		initialRouteName: Screens.Loading,
	},
);

export default createAppContainer(MainNavigator);