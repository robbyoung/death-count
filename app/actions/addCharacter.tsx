import { Death, Character } from '../state';
import { Action } from 'redux';

export interface AddCharacterAction extends Action {
    newCharacter: Character;
}

export function addCharacter(state: Character[], action: AddCharacterAction): Character[] {
    return [
        ...state,
        action.newCharacter,
    ]
}