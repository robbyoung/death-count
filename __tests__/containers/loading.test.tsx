import React from 'react';
import renderer from 'react-test-renderer';
import { StackActions, NavigationActions } from 'react-navigation';
import Loading from '../../app/containers/loading';
import { Screens } from '../../app/screens';
import { loadStateAction } from '../../app/actions';
import store from '../../app/store';
import { createTestState } from '../../app/state';

const fakeNavigation = {
    dispatch: () => undefined,
}

jest.mock('react-navigation', () => ({
    StackActions: {
        reset: jest.fn(),
    },
    NavigationActions: {
        navigate: jest.fn(),
    }
}));

jest.mock('../../app/storage', () => ({
    loadState: () => undefined,
}));

describe('Loading Container', () => {
    afterEach(() => jest.clearAllMocks());

    it('Redirects to the startup screen if there are no games', async () => {
        const component = renderer.create(<Loading navigation={fakeNavigation as any} />);
        await waitForLoad();

        expect(component.toJSON()).toMatchSnapshot();
        expect(NavigationActions.navigate).toHaveBeenCalledWith({ routeName: Screens.Startup });
        expect(StackActions.reset).toHaveBeenCalledTimes(1);
    });

    it('Redirects to the startup screen if there are only incomplete games', async () => {
        const state = createTestState(1, 0, 0, 0, 0);
        state.games[0].complete = false;
        const action = loadStateAction(state);
        store.dispatch(action);

        const component = renderer.create(<Loading navigation={fakeNavigation as any} />);
        await waitForLoad();

        expect(component.toJSON()).toMatchSnapshot();
        expect(NavigationActions.navigate).toHaveBeenCalledWith({ routeName: Screens.Startup });
        expect(StackActions.reset).toHaveBeenCalledTimes(1);
    });

    it('Redirects to the home screen if there are games', async () => {
        const action = loadStateAction(createTestState(1, 1, 0, 0, 0));
        store.dispatch(action);

        const component = renderer.create(<Loading navigation={fakeNavigation as any} />);
        await waitForLoad();

        expect(component.toJSON()).toMatchSnapshot();
        expect(NavigationActions.navigate).toHaveBeenCalledWith({ routeName: Screens.Home });
        expect(StackActions.reset).toHaveBeenCalledTimes(1);
    });
});

async function waitForLoad() {
    return new Promise(resolve => {
        setTimeout(() => resolve(), 0);
    });
}
