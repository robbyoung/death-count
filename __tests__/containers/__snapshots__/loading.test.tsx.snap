// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Loading Container Redirects to the home screen if there are games 1`] = `<View />`;

exports[`Loading Container Redirects to the startup screen if there are no games 1`] = `<View />`;

exports[`Loading Container Redirects to the startup screen if there are only incomplete games 1`] = `<View />`;
